internal/pkg/agent/application/dispatcher/dispatcher.go

```diff
@@ -192,70 +203,165 @@ func detectTypes(actions []fleetapi.Action) []string {
+	queuedUpgradeActions := maps.Collect(func(yield func(K string, V fleetapi.Action) bool) {
+		for _, action := range ad.queue.Actions() {
+			if _, ok := action.(*fleetapi.ActionUpgrade); !ok {
+				continue
+			}
+
+			if !yield(action.ID(), action) {
+				return
+			}
+		}
+	})
+
 	for i := len(actions) - 1; i >= 0; i-- {
 		action := actions[i]
-		// If it is a cancel action, remove from list and dispatch
-		if action.Type() == fleetapi.ActionTypeCancel {
-			actions = append(actions[:i], actions[i+1:]...)
-			if err := ad.dispatchAction(ctx, action, acker); err != nil {
-				ad.log.Errorf("Unable to dispatch cancel action id %s: %v", action.ID(), err)
-			}
+
+		cancelAction, ok := action.(*fleetapi.ActionCancel)
+		if !ok {
+			continue
+		}
+
+		actions = append(actions[:i], actions[i+1:]...)
+		if err := ad.dispatchAction(ctx, action, acker); err != nil {
+			ad.log.Errorf("Unable to dispatch cancel action id %s: %v", action.ID(), err)
+		}
+
+		if _, exists := queuedUpgradeActions[cancelAction.Data.TargetID]; exists {
+			*upgradeDetailsNeedUpdate = true
 		}
 	}
 	return actions
 }
 
-// gatherQueuedActions will dequeue actions from the action queue and separate those that have already expired.
-func (ad *ActionDispatcher) gatherQueuedActions(ts time.Time) (queued, expired []fleetapi.Action) {
-	actions := ad.queue.DequeueActions()
-	for _, action := range actions {
+// mergeWithQueuedActions will dequeue actions from the action queue, excluding those that have already expired, and merge
+// them with the passed actions. If upgrade actions are found to inside the dispatchable action, upgradeDetailsNeedUpdate
+// will be set to false, otherwise if upgrade actions are found to be expired, upgradeDetailsNeedUpdate will be set to true.
+func (ad *ActionDispatcher) mergeWithQueuedActions(ts time.Time, actions []fleetapi.Action, upgradeDetailsNeedUpdate *bool) []fleetapi.Action {
+	var expired []fleetapi.ScheduledAction
+	var expiredUpgradeActions []fleetapi.ScheduledAction
+	dequeuedActions := ad.queue.DequeueActions()
+
+	for _, action := range dequeuedActions {
 		exp, _ := action.Expiration()
 		if ts.After(exp) {
-			expired = append(expired, action)
+			if action.Type() == fleetapi.ActionTypeUpgrade {
+				// this is an expired upgrade action thus we need to recalculate the upgrade details
+				*upgradeDetailsNeedUpdate = true
+				expiredUpgradeActions = append(expiredUpgradeActions, action)
+			} else {
+				expired = append(expired, action)
+			}
+
+			continue
+		}
+		actions = append(actions, action)
+	}
+
+	ad.log.Debugf("Gathered %d actions from queue, %d non-upgrade actions expired, %d upgrade actions expired",
+		len(dequeuedActions), len(expired), len(expiredUpgradeActions))
+	ad.log.Debugf("Expired non-upgrade actions: %v", expired)
+	ad.log.Debugf("Expired upgrade actions (won't be removed from the queue): %v", expiredUpgradeActions)
+
+	for _, expiredUpgradeAction := range expiredUpgradeActions {
```

Comment:
Could you add a comment here explaining why we add back the expired upgrade actions to the queue?

Reply:
take a look here 6f75b29b772af12fe38091555902a056c14b9f35

Reply:
Thanks.  

Now I'm wondering why we only remove expired actions from the queue when we get a cancel or newer upgrade action removes them.  I understand we have to keep them in the queue to survive restarts so we can report that they expired in upgrade details but why not remove them immediately after that reporting? Why wait till we get a cnacel or newer upgrade action?

Anyway, I'm not going to block the improvements made by this PR on this question; it's more for my understanding.  Depending on the answer, I think it would be helpful to document it in the code somewhere but, again, that's for a follow up PR.

Reply:
no let's discuss this so all of us are in the same page and then we merge the fixes. So the reason these are added back in the queue and not removed, is to survive elastic-agent restarts; the next time agent start it will load the details from the marker or from the action queue. If expired actions are not there it will emit `nil` upgrade details so the expiration will be lost from the fleet UI. Does that explanation help?

Reply:
because if we did that and restarted this would result in our next checkin sending nil upgrade details and thus the Fleet UI wouldn't show the expiration anymore? happy to have a sync about

Reply:
Ah okay, I got it now.  We discussed this scenario in one of the team syncs.  We need the Fleet UI to keep showing the failed state from the upgrade details until the user takes some action.  Okay, makes sense now but maybe add a comment about this in the code so I don't ask this question again in 3 weeks 😅?

Reply:
I’ve added already a comment clarifying that these need to be persisted, which should be enough to prevent someone from altering this without checking in with the team first. For now, I’ll go ahead and merge this since it fixes the issue we have at hand. And if in 3 weeks you forget about it 😅 I’ll open a follow-up PR to extend the comment further and explicitly mention that this was a team decision to always show expired ones in Fleet UI until the user takes another upgrade action.